src/builtin_string.cpp

```diff
@@ -968,30 +871,32 @@ class string_replacer_t {
 
     virtual ~string_replacer_t() = default;
     int replace_count() { return total_replaced; }
-    virtual bool replace_matches(const wchar_t *arg) = 0;
+    virtual bool replace_matches(const wcstring &arg) = 0;
 };
 
 class literal_replacer_t : public string_replacer_t {
-    const wchar_t *pattern;
-    const wchar_t *replacement;
+    const wcstring pattern;
+    const wcstring replacement;
     size_t patlen;
 
    public:
-    literal_replacer_t(const wchar_t *argv0, const wchar_t *pattern_, const wchar_t *replacement_,
+    literal_replacer_t(const wchar_t *argv0, const wcstring &pattern_, const wchar_t *replacement_,
                        const options_t &opts, io_streams_t &streams)
         : string_replacer_t(argv0, opts, streams),
           pattern(pattern_),
           replacement(replacement_),
-          patlen(wcslen(pattern)) {}
+          patlen(pattern.length()) {}
 
     ~literal_replacer_t() override = default;
-    bool replace_matches(const wchar_t *arg) override;
+    bool replace_matches(const wcstring &arg) override;
 };
 
-static wcstring interpret_escapes(const wchar_t *orig) {
+static wcstring interpret_escapes(const wcstring &arg) {
     wcstring result;
 
-    while (*orig != L'\0') {
+    const wchar_t *orig = arg.c_str();
+    const wchar_t *start = arg.c_str();
+    while (orig - start < arg.size()) {
```

Comment:
I'm not too sure about stuff like this - is the size() here given in wchar_t?

Reply:
`string` and `wstring`'s `.size()` function returns the length in "underlying characters" - but is notably not unicode aware and does not include the trailing null. Note that `string` doesn't mind embedded nulls, so you could actually have a sequence of `L'\0'` in between `string.begin()` and `string.end()` (or `string.data()/string.c_str()` and `string.data()/string.c_str() + string.size()`). Edit: nevermind, I see that's the whole point of this exercise ;)

(Since `wchar_t` is 4 bytes on *nix, the unicode thing is probably not a problem.)

Reply:
Yeah, what you did here is fine